### PR TITLE
feat: fetchUserAttributes

### PR DESCRIPTION
**Pull request by haverchuck**
*Description of changes:*
Adds fetchUserAttribute API.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AuthCognito.kt [hunk 1]
```diff
@@ -447,6 +445,13 @@ public class AuthCognito : FlutterPlugin, ActivityAware, MethodCallHandler, Plug
     }
   }
 
+  fun prepareFetchAttributeResult(@NonNull flutterResult: Result, @NonNull attributes: List<AuthUserAttribute>) {
```
Comment (Amplifiyer) [inline packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AuthCognito.kt:448]: Should it be called prepareFetchAttribute**s**Result

File: packages/amplify_auth_cognito/example/ios/unit_tests/amplify_auth_cognito_tests.swift [hunk 2]
```diff
@@ -1109,6 +1107,37 @@ class amplify_auth_cognito_tests: XCTestCase {
...
+        
+        _data = [:]
+        _args = ["data": _data]
+        let call = FlutterMethodCall(methodName: "fetchUserAttributes", arguments: _args)
+        plugin.handle(call, result: {(result)->Void in
+            if let res = result as? [AuthUserAttribute] {
+                XCTAssertEqual( 0, res.count )
+            } else {
+                XCTFail()
+            }
```
Comment (Amplifiyer) [inline packages/amplify_auth_cognito/example/ios/unit_tests/amplify_auth_cognito_tests.swift:1128]: Should we not actually create or mock a userAttribute in this test?

File: packages/amplify_auth_cognito/example/lib/main.dart [hunk 3]
```diff
@@ -242,7 +251,11 @@ class _MyAppState extends State<MyApp> {
               const Padding(padding: EdgeInsets.all(10.0)),
               ElevatedButton(
                   onPressed: _getCurrentUser,
-                  child: const Text('Get CurrentUser'))
+                  child: const Text('Get CurrentUser')),
+              const Padding(padding: EdgeInsets.all(10.0)),
+              ElevatedButton(
+                  onPressed: _fetchUserAttributes,
+                  child: const Text('Get Attributes'))
```
Comment (Amplifiyer) [inline packages/amplify_auth_cognito/example/lib/main.dart:258]: nit: "Get User Attributes"

File: packages/amplify_auth_cognito/lib/method_channel_auth_cognito.dart [hunk 4]
```diff
@@ -254,6 +255,27 @@ class AmplifyAuthCognitoMethodChannel extends AmplifyAuthCognito {
     return res;
   }
 
+  @override
+  Future< List<AuthUserAttribute>> fetchUserAttributes({AuthUserAttributeRequest request}) async {
```
Comment (Amplifiyer) [inline packages/amplify_auth_cognito/lib/method_channel_auth_cognito.dart:259]: extra space in `Future< List<AuthUserAttribute>>`

File: packages/amplify_auth_cognito/lib/method_channel_auth_cognito.dart [hunk 5]
```diff
@@ -254,6 +255,27 @@ class AmplifyAuthCognitoMethodChannel extends AmplifyAuthCognito {
...
+          'data': request != null ? request.serializeAsMap() : {},
+        },
+      );
+      res = formatFetchAttributesResponse(data);
+      return res;
+    } on PlatformException catch(e) {
+      castAndThrowPlatformException(e);
+    } on Exception catch(e) {
+      throw(e);
+    }
```
Comment (Amplifiyer) [inline packages/amplify_auth_cognito/lib/method_channel_auth_cognito.dart:275]: Isn't this a no-op?

File: packages/amplify_auth_plugin_interface/lib/amplify_auth_plugin_interface.dart [hunk 6]
```diff
@@ -16,12 +16,15 @@
...
 import 'dart:async';
+import 'package:amplify_auth_plugin_interface/src/Session/AuthUserAttributeRequest.dart';
 import 'package:flutter/foundation.dart';
 import 'package:meta/meta.dart';
 import 'package:amplify_core/types/index.dart';
+import 'src/Session/AuthUserAttribute.dart';
 import 'src/types.dart';
 export 'src/types.dart';
 
+// ignore_for_file: public_member_api_docs
```
Comment (Amplifiyer) [inline packages/amplify_auth_plugin_interface/lib/amplify_auth_plugin_interface.dart:27]: We probably shouldn't add this to our interface files. All public API methods **should** have documentation.

File: packages/amplify_auth_plugin_interface/lib/src/Session/AuthUserAttribute.dart [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import 'package:flutter/foundation.dart';
+
+
+class AuthUserAttribute {
```
Comment (Amplifiyer) [inline packages/amplify_auth_plugin_interface/lib/src/Session/AuthUserAttribute.dart:19]: The naming of the files and directories are against the best practices of dart. https://dart.dev/guides/language/effective-dart/style#do-name-libraries-and-source-files-using-lowercase_with_underscores

But this is something we haven't followed throughout the library. Should we pick this up pre-release (renaming all the files)?

File: packages/amplify_auth_plugin_interface/lib/src/Session/AuthUserAttributeOptions.dart [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+class AuthUserAttributeOptions {
+
+  const AuthUserAttributeOptions();
+  Map<String, dynamic> serializeAsMap() {
+    throw UnimplementedError('serializeAsMap() has not been implemented on AuthUserAttributeOptions.');
+  }
+}
```
Comment (Amplifiyer) [inline packages/amplify_auth_plugin_interface/lib/src/Session/AuthUserAttributeOptions.dart:21]: What's the point of this empty class and part of a request signature when no data can be provided? Seems confusing.

**Review by Amplifiyer — COMMENTED** [review]
Looks good overall, some small questions below

File: packages/amplify_auth_plugin_interface/lib/src/Session/AuthUserAttributeOptions.dart (hunk 8 above)
Comment (haverchuck) [reply packages/amplify_auth_plugin_interface/lib/src/Session/AuthUserAttributeOptions.dart:21]: This is in anticipation of future, non-Cognito plugins which may have options. I believe that native libraries do this as well for this API.

File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AuthCognito.kt (hunk 1 above)
Comment (haverchuck) [reply packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AuthCognito.kt:454]: Done.

File: packages/amplify_auth_cognito/example/ios/unit_tests/amplify_auth_cognito_tests.swift (hunk 2 above)
Comment (haverchuck) [reply packages/amplify_auth_cognito/example/ios/unit_tests/amplify_auth_cognito_tests.swift:1128]: Fixed.

File: packages/amplify_auth_cognito/example/lib/main.dart (hunk 3 above)
Comment (haverchuck) [reply packages/amplify_auth_cognito/example/lib/main.dart:260]: fixed.

File: packages/amplify_auth_cognito/lib/method_channel_auth_cognito.dart (hunk 5 above)
Comment (haverchuck) [reply packages/amplify_auth_cognito/lib/method_channel_auth_cognito.dart:275]: Fixed.

File: packages/amplify_auth_plugin_interface/lib/amplify_auth_plugin_interface.dart (hunk 6 above)
Comment (haverchuck) [reply packages/amplify_auth_plugin_interface/lib/amplify_auth_plugin_interface.dart:27]: Yeah, good point.

File: packages/amplify_auth_cognito/lib/method_channel_auth_cognito.dart (hunk 4 above)
Comment (haverchuck) [reply packages/amplify_auth_cognito/lib/method_channel_auth_cognito.dart:259]: Thanks.

**Review by Amplifiyer — APPROVED** [review]

**Review by fjnoyp — APPROVED** [review]

**Pull request merged**